### PR TITLE
Manage MT-32 rendering and playback using the RW queue

### PR DESCRIPTION
**Pull request by kcgen**
This does away with the SDL mutexes, locks, and position coordination in favour of using back-pressure from the RW queue to pace the producer and consumer threads:
 - the mixer-callback thread waits until a rendered buffer is available before being fed samples
 - the rendering thread waits until space is available in the queue before rendering another buffer-full

Rendering is now performed in fixed-sized 4 KiB chunks, which is more efficient than rendering to smaller and/or per-millisecond chunks. The RW queue also performs `std::move()` operations on the rendered buffers to ensure no copies are created.

@dreamer , unfortunately I was unable to spread these changes over multiple commits; it was simply "one or the other" (without destroying compile-ability).  On the review side, I'm happy to annotate the changes in GitHub to help the review effort. 

I went ahead with this after the other PR involving benchmarking the Moody RW queue proved it to 8+ fold less-latent than a rotely implemented blocking queue. Regardless which queue we use, this design change significantly improves the code, which is where I see the most value. 

A welcome side-effect was fixing #858, which was caused by an SDL poll-wait loop experienced on some Aarch64 SBCs. Thanks to @SupervisedThinking for both reporting and testing.

### CODE REVIEW

**Comment by kcgen** [comment]
I will also run some benchmarks on the Pi and  clock-limited x86-64.

**Comment by kcgen** [comment]
x86-64 CPU-limited to 800 Mhz:

Space Quest IV configured for MT-32 music, 9000 cycles, dynamic core, and 48 KHz mixer rate. 

Release build of dosbox-staging `master` verus this branch.

Measured using `time`, which tallies all sub-threads into the `user` and `sys` durations. Launched with:
  `time /usr/src/dosbox-staging/src/dosbox sq4.bat & sleep 60 && killall dosbox`

The first five runs testing the two binaries were recorded.

**`master` scores (sorted low-to-high)**
| user | sys | duration | (user + sys) / duration |
| -- | -- | -- | -- |
| 22.631 | 1.870 | 60.155 | 40.729 |
| 22.717 | 1.923 | 60.153 | 40.962 |
| 22.934 | 1.843 | 60.181 | 41.171 |
| 22.881 | 1.976 | 60.152 | 41.324 |
| 23.339 | 1.868 | 60.231 | 41.851 |

**PR scores (sorted low-to-high)**

| user | sys | duration | (user + sys) / duration |
| -- | -- | -- | -- |
| 22.721 | 1.712 | 60.153 | 40.619 |
| 22.677 | 1.939 | 60.197 | 40.892 |
| 22.979 | 1.787 | 60.182 | 41.152 |
| 22.935 | 1.848 | 60.203 | 41.166 |
| 23.225 |  1.814 | 60.187 | 41.602 |

Overall, the numbers are a wash and at least show no performance change or regression is caused by this PR.

**Comment by kcgen** [comment]
Raspberry Pi 3. Same build and test scenario as above.

**`master` scores (sorted low-to-high)**
| user(s) | sys(s) | wall(s) | (user + sys) / wall |
| -- | -- | -- | -- |
| 60.409 | 67.263 | 2.831 | 116.032379281233 |
| 60.322 | 67.831 | 2.661 | 116.859520572925 |
| 60.288 | 68.467 | 2.265 | 117.323513800425 |
| 60.238 | 69.289 | 2.698 | 119.504299611541 |
| 60.274 | 70.352 | 1.752 | 119.627036533165 |


**PR scores (sorted low-to-high)**

| user(s) | sys(s) | wall(s) | (user + sys) / wall |
| -- | -- | -- | -- |
| 60.316 | 68.574 | 1.673 | 116.464951256715 |
| 60.269 | 68.137 | 2.22 | 116.738289999834 |
| 60.262 | 68.628 | 1.872 | 116.989147389732 |
| 60.299 | 68.345 | 2.289 | 117.139587721189 |
| 60.32 | 69.733 | 2.157 | 119.181034482759 |

These numbers are also wash, and further the proof that  no performance change or regression is caused by this PR.

**Comment by kcgen** [comment]
This PR fixes #858, the SDL MT-32 SDL poll-wait loop experienced on the PINE64 RockPro64 and Amlogic A311D boards (both of which run Aarch64 kernels and 32-bit user-spaces).

**Review by dreamer — CHANGES_REQUESTED** [review]
The first round of review. The simplification implemented in here is awesome, and the PR/commit size is just right - all the relevant changes are together, and clearly related to each other - great :)

Do you want to add to the commit message a line saying that this fixes #858? Aside of that I think we'll need a round of user testing on more commonly used systems, but I guess it will be a formality.

Looks good, let's polish this commit a bit :)

File: src/midi/midi_mt32.cpp [hunk 1]
```diff
@@ -304,64 +295,52 @@ bool MidiHandler_mt32::Open(MAYBE_UNUSED const char *conf)
...
-	SDL_LockMutex(lock.get());
-	SDL_CondSignal(framesInBufferChanged.get());
-	SDL_UnlockMutex(lock.get());
-	thread.reset();
-	lock.reset();
-	framesInBufferChanged.reset();
-
-	service->closeSynth();
-	is_open = false;
+	// Stop playback
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.cpp:318]: Makes sense, and will do.
Do you think I should use it in place of the `continue_rendering` bool? 

The code now only needs a single `bool` to control MT-32 state (be it called open/close or rendering/not-rendering)?

File: src/midi/midi_mt32.h [hunk 2]
```diff
@@ -26,32 +26,33 @@
...
 #define MT32EMU_API_TYPE 3
 #include <mt32emu/mt32emu.h>
 #if MT32EMU_VERSION_MAJOR != 2 || MT32EMU_VERSION_MINOR < 1
 #error Incompatible mt32emu library version
 #endif
 
-#include <SDL_thread.h>
-
 #include "mixer.h"
+#include "../libs/readerwriterqueue/readerwritercircularbuffer.h"
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h:40]: I'm all for it. 
I'll make this the first commit in this PR; given it's really the first user of the RW-queue anyway.

File: src/midi/midi_mt32.h [hunk 3]
```diff
@@ -26,32 +26,33 @@
...
 
 class MidiHandler_mt32 final : public MidiHandler {
 private:
-	// Scoped types
+	static constexpr int FRAMES_PER_BUFFER = 2048; // synth granularity
+	static constexpr int SAMPLES_PER_BUFFER = FRAMES_PER_BUFFER * 2; // L & R
```
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h:45]: Good catch; I'll try dropping the `static`; certainly nothing static needed about these two variables.
Will also move the lines as you've suggested.

File: src/midi/midi_mt32.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h:40]: Renaming in-progress on a separate branch, and will push to `master` when ready. 
Then I'll update the two open PRs that use the MQ.

File: src/midi/midi_mt32.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h:45]: >  is there any difference between static constexpr int and constexpr int class member field?

``` text
Making all in midi
  CXX      midi.o
In file included from midi.cpp:79:
midi_mt32.h:44:2: error: non-static data member ‘FRAMES_PER_BUFFER’ declared ‘constexpr’
   44 |  constexpr int FRAMES_PER_BUFFER = 2048; // synth granularity
      |  ^~~~~~~~~
midi_mt32.h:45:2: error: non-static data member ‘SAMPLES_PER_BUFFER’ declared ‘constexpr’
   45 |  constexpr int SAMPLES_PER_BUFFER = FRAMES_PER_BUFFER * 2; // L & R
      |  ^~~~~~~~~
midi_mt32.h:47:39: error: invalid use of non-static data member ‘MidiHandler_mt32::SAMPLES_PER_BUFFER’
   47 |  using buffer_t = std::array<int16_t, SAMPLES_PER_BUFFER>;
      |                                       ^~~~~~~~~~~~~~~~~~
```
(keeping `static`)

File: src/midi/midi_mt32.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h]: Rename done and PR rebased.

File: src/midi/midi_mt32.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (kcgen) [reply src/midi/midi_mt32.h:45]: >  please move `using buffer_t` line below or `using ring_t` line above - so those two lines would be next to each other (I want it to be easier to notice that `ring_t` depends on `buffer_t`).

Nice suggestion and the connection is now apparent.

File: src/midi/midi_mt32.cpp [hunk 4]
```diff
@@ -378,98 +367,40 @@ void MidiHandler_mt32::PlaySysex(uint8_t *sysex, size_t len)
 	service->playSysexAt(sysex, msg_len, GetMidiEventTimestamp());
 }
 
-int MidiHandler_mt32::ProcessingThread(MAYBE_UNUSED void *data)
+// Ensures the member 'buffer' still has unplayed frames, and returns the quantity
+uint16_t MidiHandler_mt32::PopBuffer()
```
Comment (dreamer) [inline src/midi/midi_mt32.cpp:371]: This function name is really confusing considering the comment before. When reading code I kept thinking that PopBuffer does what `pop` operation usually does: returns the value from buffer (not number of elements).

Why comment in this function refer to "next buffer"?

Also, I think member variable `played_buffers` is confusingly named (shouldn't it be `played_frames` or something?).

**Review by dreamer — CHANGES_REQUESTED** [review]
OK, I left comments about some stylistic/naming issues. Aside of that, I am ok with merging it in.

When thinking how to test this I remembered that PakuPaku can use MT-32 for sound effects - @kcgen can you verify if new implementation works well in this game? To run MT-32 sound effects `paku.exe /mt32`.

File: src/midi/midi_mt32.cpp (hunk 4 above)
Comment (dreamer) [reply src/midi/midi_mt32.cpp:371]: Suggestion: maybe this method should named something like `GetBufferredFramesNumber`?

Comment (kcgen) [reply src/midi/midi_mt32.cpp:371]: Thanks @dreamer. I will work in re-naming these and adding comments.

> Also, I think member variable **played_buffers** is confusingly named (shouldn't it be **played_frames**?)

The ring operates on "full buffer" objects. It actually holds three distinct `buffer_t` objects (each of  which inside holds 2048 frames).

Playback (in the mixer callback) happens at the frame-level. We steadily chew through the requested numbers of frames (in the current popped buffer), until we've played all its frames, and then "pop" out the next buffer to give us more frames to play (if a buffer isn't available then we're block and wait for the ring to give us one).

The moment we pop it out a new buffer for playing, back-pressure is released in the ring and MT-32 is unblocked to start rendering a new "full buffer", which is added to the ring.

We need to keep track of our actual played position, so this is done with two counters:
 - total "full buffers" played
 - plus the current frame played in the current buffer
 - The two are combined: `played_buffers * FRAMES_PER_BUFFER + current_frame`, to give us the total  frames played, which MT-32 uses to compute a time offset.

But yeah - these aren't obvious enough and I'll work on making it clearer.

Comment (kcgen) [reply src/midi/midi_mt32.cpp]: @dreamer , see the changes in: https://github.com/dosbox-staging/dosbox-staging/pull/863/commits/71d16408c848185d3d872ee0673997a477f3b6bf

Once we're done, I will squash this (and any additional fixes) down into the original commit.

**Comment by kcgen** [comment]
> can you verify if new implementation works well in this game? To run MT-32 sound effects paku.exe /mt32.

[paku_000.flac.zip](https://github.com/dosbox-staging/dosbox-staging/files/5939770/paku_000.flac.zip)

``` text
dosbox-staging version git
---
CONFIG: Loading primary file dosbox.conf
OPENGL: Pixel buffer object extension: available
SDL: Mouse will be captured immediately on start and middle-click will uncapture the mouse.
MEMORY: Base address: 0x7f6209666010
MEMORY: Using 256 DOS memory pages (1 MiB)
MIXER: Negotiated 2-channel 48000-Hz audio in 1024-frame blocks
MT32: Found ROM pair in ../roms/
MT32: Initialized CM-32L/LAPC-I Control v1.02 with CM-32L/CM-64/LAPC-I PCM ROM
MIXER: MT32 channel operating at 48000 Hz without resampling
MIDI: Opened device: mt32
MAPPER: Found no joysticks
MAPPER: Loaded default key bindings
MOUNT: Path '.' found
SDL: captured the mouse
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
MAIN: Pixel-perfect scaling (1x1): 640x400 (PAR 1.20) -> 640x400 (PAR 1.00)
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
MT32: LCD-Message:      PAKU  PAKU     
MT32: LCD-Message: ***  PAKU  PAKU  ***
```

File: src/midi/midi_mt32.cpp (hunk 4 above)
Comment (dreamer) [reply src/midi/midi_mt32.cpp]: Ehh, I read these new changes and I still don't fully understand this algorithm… Why do we need to use a circular buffer to store 3 array buffers storing `int16_t` samples instead of simply storing the samples directly in the circular buffer? Isn't this an unnecessary complication? Am I missing something?

I guess maybe old code used 3 buffers, so mutex operations could be avoided in practice… but if we have fancy circular buffer that can be used from multiple threads at the same time, then having 3 buffers is pointless, I think?

Comment (kcgen) [reply src/midi/midi_mt32.cpp]: Yes, we could enqueue and dequeue individual samples directly, but then back pressure on the rendered would then paced per-sample: literally individually, and we don't want to be rendering individual samples because it's extremely inefficient.

Even operating on 1ms block-sizes (every 44 or 48 frames), is still extremely inefficient.

The old code also rendered in the same large chunk sizes, but didn't have the queue controlling the threads - so that's why the book keeping and thread control was so nasty.  

The win (in this bit of code) is the blocking queue being able to pace the threads, as opposed to using microscopic per-sample chunks (which I would do, if it were efficient... But like most things in audio they like to read or write to  cache-line-sized chunks of data). 

When it comes to the modem FIFO: hardware could read and write individual bytes simultaneously, but DOSBox doesn't because the exists FIFO s aren't thread safe. So that's another area where this queue can help and we will absolutely need to operate per-byte 👍

Comment (kcgen) [reply src/midi/midi_mt32.cpp]: We're eating through a full queue every 42 ms, at which point the MT32 is asked to start rendering another buffer-worth. 

It's not a lot of work for the queue or the threads.. which is a good thing to keep their involvement as minimal as possible given the MT32 is very CPU intensive (needing roughly 900 to 1000 MHz; it's too much to even run on the Mister FPGA and instead one full Pi3 core is essentially consumed).

To put it another way: if someone submitted a PR to render individual samples, my first comment would be to chunk it to 2 or 4 KiB :-)

**Review by dreamer — APPROVED** [review]
OK, I think my worries were sufficiently explained-away :)

I'll merge the code after the last commit will be squashed however you find it appropriate.

About PakuPaku - but what's the result of the test? I can't test it with MT-32 support turned on due to no ROMs, but when I test with FluidSynth, then sound effects are rendered practically immediately (e.g. when Pacman eats a pellet). Does it work equally well with MT-32? (No matter what will be the result - if the sound is playing and there's no regression, then it's ok to merge the code in).

**Comment by kcgen** [comment]
Squashed and inbound; thanks for the thorough review @dreamer .

> About PakuPaku - but what's the result of the test?

I visually can't see any latency differences between all three (MT32-master, MT32-RW-queue, MIDI).  

The dealth animation and symbol crash all appear when the little "X" vanishes with the symbols trailing.

[paku-side-by-side.zip](https://github.com/dosbox-staging/dosbox-staging/files/5940311/paku-side-by-side.zip)

If we want to, we can ratchet latency even tighter by shortening the buffer frames (from 2048 to to say 1024 or even 512), but then we'd only be rendering one or two frames in advance, which is a bit too close for comfort (atleast on systems like the Pi).

**Comment by dreamer** [comment]
@kcgen Thanks! :)
